consensus/istanbul/backend/engine_test.go

```diff
@@ -0,0 +1,507 @@
+		t.Errorf("error mismatch: have %v, want %v", err, errInvalidExtraDataFormat)
+	}
+
+	// invalid difficulty
+	block = makeBlockWithoutSeal(chain, engine, chain.Genesis())
+	header = block.Header()
+	header.BlockScore = big.NewInt(2)
+	err = engine.VerifyHeader(chain, header, false)
+	if err != errInvalidBlockScore {
+		t.Errorf("error mismatch: have %v, want %v", err, errInvalidBlockScore)
+	}
+
+	// invalid timestamp
+	block = makeBlockWithoutSeal(chain, engine, chain.Genesis())
+	header = block.Header()
+	header.Time = new(big.Int).Add(chain.Genesis().Time(), new(big.Int).SetUint64(engine.config.BlockPeriod-1))
+	err = engine.VerifyHeader(chain, header, false)
+	if err != errInvalidTimestamp {
+		t.Errorf("error mismatch: have %v, want %v", err, errInvalidTimestamp)
+	}
+
+	// future block
+	block = makeBlockWithoutSeal(chain, engine, chain.Genesis())
+	header = block.Header()
+	header.Time = new(big.Int).Add(big.NewInt(now().Unix()), new(big.Int).SetUint64(10))
+	err = engine.VerifyHeader(chain, header, false)
+	if err != consensus.ErrFutureBlock {
+		t.Errorf("error mismatch: have %v, want %v", err, consensus.ErrFutureBlock)
+	}
+
+	// TODO-Klaytn: add more tests for header.Governance, header.Rewardbase, header.Vote
+}
+
+func TestVerifySeal(t *testing.T) {
+	chain, engine := newBlockChain(1)
+	defer engine.Stop()
+
+	genesis := chain.Genesis()
+
+	// cannot verify genesis
+	err := engine.VerifySeal(chain, genesis.Header())
+	if err != errUnknownBlock {
+		t.Errorf("error mismatch: have %v, want %v", err, errUnknownBlock)
+	}
+	block := makeBlock(chain, engine, genesis)
+
+	// change block content
+	header := block.Header()
+	header.Number = big.NewInt(4)
+	block1 := block.WithSeal(header)
+	err = engine.VerifySeal(chain, block1.Header())
+	if err != errUnauthorized {
+		t.Errorf("error mismatch: have %v, want %v", err, errUnauthorized)
+	}
+
+	// unauthorized users but still can get correct signer address
+	engine.privateKey, _ = crypto.GenerateKey()
+	err = engine.VerifySeal(chain, block.Header())
+	if err != nil {
+		t.Errorf("error mismatch: have %v, want nil", err)
+	}
+}
+
+func TestVerifyHeaders(t *testing.T) {
+	chain, engine := newBlockChain(1)
+	defer engine.Stop()
+
+	genesis := chain.Genesis()
+
+	// success case
+	headers := []*types.Header{}
+	blocks := []*types.Block{}
+	size := 100
+
+	for i := 0; i < size; i++ {
+		var b *types.Block
+		if i == 0 {
+			b = makeBlockWithoutSeal(chain, engine, genesis)
+			b, _ = engine.updateBlock(genesis.Header(), b)
+			engine.db.WriteHeader(b.Header())
+		} else {
+			b = makeBlockWithoutSeal(chain, engine, blocks[i-1])
+			b, _ = engine.updateBlock(blocks[i-1].Header(), b)
+			engine.db.WriteHeader(b.Header())
+		}
+		blocks = append(blocks, b)
+		headers = append(headers, blocks[i].Header())
+	}
+
+	// proceed time to avoid future block errors
+	now = func() time.Time {
+		return time.Unix(headers[size-1].Time.Int64(), 0)
+	}
+	defer func() {
+		now = time.Now
+	}()
```

Comment:
This defer code is necessary?

Reply:
Yes. When the test case is executed with other test cases in a sequence, modified `now` could cause unexpected test failures.

Reply:
I got it. :)